apps/emqx_retainer/src/emqx_retainer_mnesia.erl

```diff
@@ -258,4 +381,172 @@ is_table_full() ->
+
+            %% Clear old index records.
+            {atomic, ok} = mria:clear_table(?TAB_INDEX),
+
+            %% Fill index records in batches.
+            QH = qlc:q([Topic || #retained_message{topic = Topic} <- ets:table(?TAB_MESSAGE)]),
+            ok = reindex_batch(qlc:cursor(QH), 0, StatusFun),
+
+            %% Enable read indices and unlock reindexing.
+            finalize_reindex();
+        {atomic, Reason} ->
+            Reason
+    end.
+
+try_start_reindex(NewIndices, true) ->
+    mria:transaction(
+      ?RETAINER_SHARD,
+      fun() -> start_reindex(NewIndices) end);
+
+try_start_reindex(NewIndices, false) ->
+    mria:transaction(
+      ?RETAINER_SHARD,
+      fun() ->
+              case mnesia:read(?TAB_INDEX_META, ?META_KEY, write) of
+                  [#retained_index_meta{reindexing = false}] ->
+                      start_reindex(NewIndices);
+                  _ -> {error, already_started}
+              end
+      end).
+
+start_reindex(NewIndices) ->
+    ?tp(warning, retainer_message_reindexing_started, #{}),
+    mnesia:write(
+      ?TAB_INDEX_META,
+      #retained_index_meta{
+         key = ?META_KEY,
+         read_indices = [],
+         write_indices = NewIndices,
+         reindexing = true},
+      write).
+
+finalize_reindex() ->
+    {atomic, ok} = mria:transaction(
+                    ?RETAINER_SHARD,
+                    fun() ->
+                            case mnesia:read(?TAB_INDEX_META, ?META_KEY, write) of
+                                [#retained_index_meta{write_indices = WriteIndices} = Meta] ->
+                                    mnesia:write(
+                                      ?TAB_INDEX_META,
+                                      Meta#retained_index_meta{
+                                        key = ?META_KEY,
+                                        read_indices = WriteIndices,
+                                        reindexing = false},
+                                      write);
+                                [] -> ok
+                            end
+                    end),
+    ?tp(warning, retainer_message_reindexing_finished, #{}),
+    ok.
+
+reindex_topic(Indices, Topic) ->
+    case mnesia:read(?TAB_MESSAGE, Topic, read) of
+        [#retained_message{expiry_time = ExpiryTime}] ->
+            ok = emqx_retainer_index:foreach_index_key(
+                   fun(Key) -> store_retained_index(Key, ExpiryTime) end,
+                   Indices,
+                   Topic);
+        [] -> ok
+    end.
+
+reindex_batch(QC, Done, StatusFun) ->
+    Fun = fun() ->
+                  Indices = db_indices(write),
+                  {Status, Topics} = qlc_next_answers(QC, ?REINDEX_BATCH_SIZE),
+                  ok = lists:foreach(
+                    fun(Topic) -> reindex_topic(Indices, Topic) end,
+                    Topics),
+                  {Status, Done + length(Topics)}
+          end,
+    case mria:transaction(?RETAINER_SHARD, Fun) of
```

Comment:
Q: if someone issues a force reindex while a reindex is in progress, won't there be 2 processes writing to the tables?  There could be records still being inserted into a deleted index (assuming the user changed the index config) by the old reindexing process.

Perhaps, as a possible workaround, we could insert some `erlang:make_ref/0` into the `?TAB_INDEX_META` record and pass that to `reindex_batch/4`, so it can check if the reference is the same as expected.  If there's a mismatch, that would indicate that a new reindex has started and this one should stop.

Does this make sense?

Reply:
New reindex will transactionally update write index specs and then clear index table. Each batch rereads actual index specs in the beginning of each batch, so the old and new reindex processes will use the same new indices.
Some writes will be duplicated, but the result will be still consistent.